src/main/java/net/earthcomputer/clientcommands/command/ListenCommand.java

```diff
@@ -0,0 +1,269 @@
+                } else {
+                    try {
+                        packetDataPreview = serialize(packet);
+                        packetData = packetDataPreview.getString();
+                    } catch (StackOverflowError e) {
+                        e.printStackTrace();
+                        return;
+                    }
+                }
+
+                MutableComponent packetComponent = Component.literal(mojmapPacketName.orElseThrow().substring(MOJMAP_PACKET_PREFIX.length())).withStyle(s -> s
+                    .withUnderlined(true)
+                    .withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, packetDataPreview))
+                    .withClickEvent(new ClickEvent(ClickEvent.Action.COPY_TO_CLIPBOARD, packetData)));
+
+                switch (side) {
+                    case CLIENTBOUND -> source.sendFeedback(Component.translatable("commands.clisten.receivedPacket", packetComponent));
+                    case SERVERBOUND -> source.sendFeedback(Component.translatable("commands.clisten.sentPacket", packetComponent));
+                }
+            };
+        }
+
+        return Command.SINGLE_SUCCESS;
+    }
+
+    private static int remove(FabricClientCommandSource source, Class<Packet<?>> packetClass) throws CommandSyntaxException {
+        if (!packets.remove(packetClass)) {
+            throw NOT_LISTENING_EXCEPTION.create();
+        }
+
+        source.sendFeedback(Component.translatable("commands.clisten.remove.success"));
+        return Command.SINGLE_SUCCESS;
+    }
+
+    private static int list(FabricClientCommandSource source) {
+        int amount = packets.size();
+        if (amount == 0) {
+            source.sendFeedback(Component.translatable("commands.clisten.list.none"));
+        } else {
+            source.sendFeedback(Component.translatable("commands.clisten.list"));
+            packets.forEach(packetClass -> {
+                String packetClassName = packetClass.getName().replace('.', '/');
+                Optional<String> mojmapName = MappingsHelper.namedOrIntermediaryToMojmap_class(packetClassName);
+                source.sendFeedback(Component.literal(mojmapName.orElseThrow().substring(MOJMAP_PACKET_PREFIX.length())));
+            });
+        }
+
+        return amount;
+    }
+
+    private static int clear(FabricClientCommandSource source) {
+        int amount = packets.size();
+        packets.clear();
+        source.sendFeedback(Component.translatable("commands.clisten.clear"));
+        return amount;
+    }
+
+    private static Component serialize(Object object) {
+        try {
+            if (SEEN.get().add(object)) {
+                return serializeInner(object);
+            }
+            return Component.empty();
+        } finally {
+            SEEN.get().remove(object);
+            if (SEEN.get().isEmpty()) {
+                SEEN.remove();
+            }
+        }
+    }
+
+    private static Component serializeInner(Object object) {
```

Comment:
I am worried that Mojang might start using a different JDK class that we might want to pretty print (like you already had to add `Date` and `Instant`). This was the maintenance burden I was referring to in Discord. But at least we gracefully fallback to listing the fields in this case.

Reply:
I added `Date` only because `Component#translationArg(Date)` exists and I added `Instant` because the result looks nicer as a `Date`. I don't think JDK classes are a problem anymore with this `IMPL_LOOKUP` solution.